### PR TITLE
Model adapter

### PR DESCRIPTION
**Pull request by sgalkina**
Constrain the model according to:
- genotype changes
- medium definition
- experimental measurements

### CODE REVIEW

**Comment by sgalkina** [comment]
The fix to `gnomic` library is needed, there is a line https://github.com/biosustain/gnomic-python/blob/c44c9d7757d8fd91a7f248182421581dd17b10a4/gnomic/models.py#L91 which is incompatible to Python 2.7

File: driven/generic/adapter.py [hunk 1]
```diff
@@ -175,18 +181,250 @@ def annotate_new_metabolite(metabolite):
...
+
+    def apply_mutation(self, mutation):
+        """Apply mutations on initial model
+
+        :param mutation: gnomic.Mutation
+        :return:
+        """
+        if mutation.old:
+            for feature in mutation.old.features():
+                self.knockout_gene(feature)
```
Comment (joaocardoso) [inline driven/generic/adapter.py]: Is the assumption here that mutations will always knockout the gene? Shouldn't it say something about that in the documentation? What about mutations that change the activity, like constraining it to half?

File: driven/generic/adapter.py [hunk 2]
```diff
@@ -175,18 +181,250 @@ def annotate_new_metabolite(metabolite):
...
+            return
+        for reaction_id, equation in self.genes_to_reactions.get(identifier, {}).items():
+            self.add_reaction(reaction_id, equation, identifier)
+        logger.info('Gene added: {}'.format(identifier))
+
+    def add_reaction(self, reaction_id, equation, gene_name):
+        """Add new reaction by rn ID from equation, where metabolites defined by kegg ids.
+
+        :param reaction_id: reaction rn ID
+        :param equation: equation string, where metabolites are defined by kegg ids
```
Comment (joaocardoso) [inline driven/generic/adapter.py]: Are KEGG ids the only valid ids? If yes, you can make a check before hand. KEGG ids match the expression /C\d{5}/

File: driven/generic/adapter.py [hunk 3]
```diff
@@ -175,18 +181,250 @@ def annotate_new_metabolite(metabolite):
...
+    Applies medium on cameo model
+    """
+
+    def __init__(self, model, medium):
+        """
+        Parameters
+        ----------
+        model
+            cameo model
+        medium
```
Comment (joaocardoso) [inline driven/generic/adapter.py]: The medium change model can set the correct uptake rates if you know them. It is probably more relevant than the concentration. You can add an extra key to the dictionaries in the list.

File: driven/generic/adapter.py [hunk 4]
```diff
@@ -175,18 +181,250 @@ def annotate_new_metabolite(metabolite):
...
+        for compound in self.medium:
+            existing_metabolite = self.model_metabolite(compound['id'], '_e')
+            if not existing_metabolite:
+                logger.info('No metabolite {}'.format(compound['id']))
+                continue
+            logger.info('Found metabolite {}'.format(compound['id']))
+            self.add_demand_reaction(existing_metabolite)
+
+
+class MeasurementChangeModel(ModelModificationMixin):
```
Comment (joaocardoso) [inline driven/generic/adapter.py]: See https://github.com/biosustain/driven/blob/master/driven/flux_analysis/fluxomics.py and https://github.com/biosustain/driven/blob/master/driven/data_sets/fluxes.py

Flux measurements are a data type like expression or concentrations.

**Pull request closed without merging**